Ghidra/Processors/STM/data/languages/ST7.sinc

```diff
@@ -0,0 +1,758 @@
+BITOP: ([imm8],Y)       is $(PRE_PIY) & addressing=0x06 & Y ; imm8  { local indirect:2 = imm8; local address:1 = *:1 indirect + zext(Y); export *:1 address; }
+
+REL: reloc              is rel [ reloc = inst_next + rel; ]         { export *:2 reloc; }
+
+JCC: "C"                is opcode=0x25                              { export C; }
+JCC: "EQ"               is opcode=0x27                              { export Z; }
+JCC: "F"                is opcode=0x21                              { export 0:1; }
+JCC: "H"                is opcode=0x29                              { export H; }
+JCC: "IH"               is opcode=0x2F                              { export EXTI; }
+JCC: "IL"               is opcode=0x2E                              { local tmp = !EXTI; export tmp; }
+@ifdef ST70
+JCC: "M"                is opcode=0x2D                              { export I; }
+@endif
+@ifdef ST72
+JCC: "M"                is opcode=0x2D                              { local tmp = ((I1 << 1) | I0) == 0b11; export tmp; }
+@endif
+JCC: "MI"               is opcode=0x2B                              { export N; }
+JCC: "NC"               is opcode=0x24                              { local tmp = !C; export tmp; }
+JCC: "NE"               is opcode=0x26                              { local tmp = !Z; export tmp; }
+JCC: "NH"               is opcode=0x28                              { local tmp = !H; export tmp; }
+@ifdef ST70
+JCC: "NM"               is opcode=0x2C                              { local tmp = !I; export tmp; }
+@endif
+@ifdef ST72
+JCC: "NM"               is opcode=0x2C                              { local tmp = ((I1 << 1) | I0) == 0; export tmp; }
+@endif
+JCC: "PL"               is opcode=0x2A                              { local tmp = !N; export tmp; }
+JCC: "UGT"              is opcode=0x22                              { local tmp = !(C | Z); export tmp; }
+JCC: "ULE"              is opcode=0x23                              { local tmp = (C | Z); export tmp; }
+
+macro nzFlags(value) {
+  Z = value == 0;
+  N = value s< 0;
+}
+
+:ADC A, ASMEMOP is kind=0x09 ... & ASMEMOP & A {
+  local value = ASMEMOP;
+  local oldCarry = C;
+  C = carry(A, value);
+  H = carry(A << 4, value << 4);
+  value = value + A + oldCarry;
+  nzFlags(value);
+  A = value;
+}
+
+:ADD A, ASMEMOP is kind=0x0B ... & ASMEMOP & A {
+  local value = ASMEMOP;
+  C = carry(A, value);
+  H = carry(A << 4, value << 4);
+  value = value + A;
+  nzFlags(value);
+  A = value;
+}
+
+:AND A, ASMEMOP is kind=0x04 ... & ASMEMOP & A {
+  local value = ASMEMOP;
+  value = value & A;
+  nzFlags(value);
+  A = value;
+}
+
+:BCP A, ASMEMOP is kind=0x05 ... & ASMEMOP & A {
+  local value = ASMEMOP;
+  local result = value & A;
+  nzFlags(result);
+}
+
+:BRES imm8, "#"bit_position is $(NO_PRE) & bit_upper=0x01 & bit_lower=0x01 & bit_position ; imm8 {
```

Comment:
maybe not worth worrying about.  wonder if you could use a table to remove some of this duplication, untested but then all these instructions would just be something like this... i think this is possible.

```
PIX: imm8 is $(NO_PRE) & imm8 { local address:2 = imm8; export address; }
PIX: [imm8] is $(PRE_PIX) & imm8 { local address:2 = imm8; local target = *:1 address; export target; }

:BRES PIX, "#"bit_position is bit_upper=0x01 & bit_lower=0x01 & bit_position ; PIX {
  local value:1 = *:1 PIX;
  value = value & ~(1 << bit_position);
  *:1 PIX = value;
}
```

Reply:
This doesn't seem to work:

`ST7.sinc:121: Table 'PIX' has inconsistent export size; Constructor at ST7.sinc:121 is first conflict`

Line 121 is the [imm8] definition.  I'm looking into this...

Reply:
Nope, it just gets confused and fails to decode opcodes.  I guess I'll keep two copies of each opcode for now :(

Reply:
Hrm, if I apply the same approach to JR^JCC it works. There must be something wrong going on with prefixes.